src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -37,19 +41,25 @@ public override async Task<Compilation> TransformCompilationAsync(Compilation ol
                 public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generatorDriver)
                     => generatorDriver;
 
-                public override CompilationAndGeneratorDriverTranslationAction? TryMergeWithPrior(CompilationAndGeneratorDriverTranslationAction priorAction)
+                public override TranslationAction? TryMergeWithPrior(TranslationAction priorAction)
                 {
                     if (priorAction is TouchDocumentAction priorTouchAction &&
                         priorTouchAction._newState == _oldState)
                     {
-                        return new TouchDocumentAction(priorTouchAction._oldState, _newState);
+                        // As we're merging ourselves with the prior touch action, we want to keep the old project state
+                        // that we are translating from.
+                        return new TouchDocumentAction(priorAction.OldProjectState, priorTouchAction._oldState, _newState);
```

Comment:
this preserves the sematnics that after merging, our 'oldstate' was the original old state of the prior action.